### ISSUE TITLE
Possible SQL query optimization for retrival from the tags table

### ISSUE THREAD

**Issue by lapin-b** [diaspora/diaspora#8210]
# The context

Poking around Diaspora's database and trying to figure out what's happening with slow queries, I stumbled across a trivial query that takes 5 to 7 seconds to return:

```sql
SELECT "tags".* FROM "tags" WHERE (LOWER(name) = LOWER('mintpressnews') OR LOWER(name) = LOWER('dakota') OR LOWER(name) = LOWER('minnesota'));
```

The query plan looks like as follows:

```                                                                                        
                                                                       QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.00..18835.22 rows=13826 width=21) (actual time=914.236..925.473 rows=3 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Parallel Seq Scan on tags  (cost=0.00..16452.62 rows=5761 width=21) (actual time=388.894..907.331 rows=1 loops=3)
         Filter: ((lower((name)::text) = 'mintpressnews'::text) OR (lower((name)::text) = 'dakota'::text) OR (lower((name)::text) = 'minnesota'::text))
         Rows Removed by Filter: 308787
 Planning Time: 0.258 ms
 Execution Time: 925.549 ms
```
(execution time is less than one second, the data was very recently fetched)

The only way the planner found to execute the query is to do a parallel sequential scan on the table. Except the scan is on a table with close to one million rows (according to `SELECT COUNT(*) FROM tags`) and is slow, especially when the load on the server increases

# Solution

There is one solution which doesn't need any database structure modification: editing the SQL query to no longer explicitly lower the column and value:

```sql
SELECT "tags".* FROM "tags" WHERE name = 'mintpressnews' OR name = 'dakota' OR name = 'minnesota'
```

This enables PostgreSQL to use the index and do bitmap operations on it, which is already way faster than a sequential scan. Note: `LOWER(...)` can still be applied, but only on the right side of the equality operator. Resulting query plan:

```
                                                            QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on tags  (cost=13.30..25.16 rows=3 width=21) (actual time=2.596..2.605 rows=3 loops=1)
   Recheck Cond: (((name)::text = 'mintpressnews'::text) OR ((name)::text = 'dakota'::text) OR ((name)::text = 'minnesota'::text))
   Heap Blocks: exact=3
   ->  BitmapOr  (cost=13.30..13.30 rows=3 width=0) (actual time=2.577..2.577 rows=0 loops=1)
         ->  Bitmap Index Scan on index_tags_on_name  (cost=0.00..4.43 rows=1 width=0) (actual time=0.829..0.829 rows=1 loops=1)
               Index Cond: ((name)::text = 'mintpressnews'::text)
         ->  Bitmap Index Scan on index_tags_on_name  (cost=0.00..4.43 rows=1 width=0) (actual time=1.184..1.184 rows=1 loops=1)
               Index Cond: ((name)::text = 'dakota'::text)
         ->  Bitmap Index Scan on index_tags_on_name  (cost=0.00..4.43 rows=1 width=0) (actual time=0.559..0.559 rows=1 loops=1)
               Index Cond: ((name)::text = 'minnesota'::text)
 Planning Time: 0.382 ms
 Execution Time: 2.694 ms

```
(the execution time is in milliseconds, the data is still "warm" from previous queries)

Note the bitmap index scan operation and the resulting execution time which is much better than 

Of course, we will need to explicitly make the tags lowercase in the Ruby code before selecting, inserting or updating any tags in the table. Furthermore, we will have to ensure that all the tags present in the table are lowercase with an update in a migration and the index already exists.

**Comment by Flaburgan** [comment]
Hello @lapin-b and thank you for taking time to investigate that problem!
We already saw it and merged a patch (see https://github.com/diaspora/diaspora/pull/8206) which is similar to your proposition. It is not released yet but will be part of the coming 0.8.0.0 release. Can you please confirm that it is indeed the same problem?

**Comment by lapin-b** [comment]
The PR you have mentioned solves one part of the problem: storing lower case values.

The other part of the issue is the SQL query issued by the application which prevents the query optimizer to use indexes. Transformation functions such as `LOWER` on the value side of the comparison seems fine, but not on the column side.

I can't really provide where exactly despite having looked around the code a little bit, since I just skimmed over to hopefully find something interesting (`TagsController` and called methods) , unless I looked in the wrong place.

**Comment by denschub** [comment]
We cannot edit the query as we're not the ones generating it - that query is generated by a dependency (ActsAsTaggableOn). However, to address the exact performance issue you ran into, we built the index that Fla already linked to. This new index is not on `name`, but on `lower(name)`. Postgres 13s planner picked those up just fine to resolve those `LOWER(name) = LOWER('foo')` queries - we tested that. MySQL is out of luck, but there's really nothing we can do for MySQL.
